File: 04-string-compression/solution.py
# String Compression
# Return compressed string only if it's shorter than original
# Time: O(n), Space: O(n)

class Solution:
    def compress(self, s: str) -> str:
        if not s:
            return s
        
        compressed = []
        count = 1
        
        for i in range(1, len(s) + 1):
            if i < len(s) and s[i] == s[i - 1]:
                count += 1
            else:
                compressed.append(s[i - 1] + str(count))
                count = 1
        
        result = ''.join(compressed)
        # Return original if compressed is not shorter
        return result if len(result) < len(s) else s

# Test
if __name__ == "__main__":
    sol = Solution()
    print(sol.compress("aaabbc"))   # Output: a3b2c1
    print(sol.compress("abc"))      # Output: abc (original)
    print(sol.compress("aabbcc"))   # Output: aabbcc (original)

File: 04-string-compression/solution.ts
// String Compression
// Return compressed string only if it's shorter than original
// Time: O(n), Space: O(n)

class Solution {
    compress(s: string): string {
        if (!s) {
            return s;
        }
        
        const compressed: string[] = [];
        let count = 1;
        
        for (let i = 1; i <= s.length; i++) {
            if (i < s.length && s[i] === s[i - 1]) {
                count++;
            } else {
                compressed.push(s[i - 1] + count);
                count = 1;
            }
        }
        
        const result = compressed.join('');
        // Return original if compressed is not shorter
        return result.length < s.length ? result : s;
    }
}

// Test
if (require.main === module) {
    const sol = new Solution();
    console.log(sol.compress("aaabbc"));   // Output: a3b2c1
    console.log(sol.compress("abc"));      // Output: abc (original)
    console.log(sol.compress("aabbcc"));   // Output: aabbcc (original)
}

export default Solution;
